File: Disk Scheduling java/cscan.java
import java.util.*;

class cscan8{

	static int disk_size;

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		System.out.print("Enter the number of requests: ");
		int size = sc.nextInt();
		int[] arr = new int[size];
		System.out.println("Enter the requests:");
		for (int i = 0; i < size; i++) {
			arr[i] = sc.nextInt();
		}

		System.out.print("Enter the initial head position: ");
		int head = sc.nextInt();

		System.out.print("Enter the disk size: ");
		disk_size = sc.nextInt();

		System.out.print("Enter the direction of head movement (left or right): ");
		String direction = sc.next();
		sc.close();

		System.out.println("Initial position of head: " + head);

		CSCAN(arr, head, direction);
	}

	public static void CSCAN(int arr[], int head, String direction) {
		int seek_count = 0;
		int distance, cur_track;
		Vector<Integer> left = new Vector<Integer>();
		Vector<Integer> right = new Vector<Integer>();
		Vector<Integer> seek_sequence = new Vector<Integer>();

		left.add(0);
		right.add(disk_size - 1);

		for (int i = 0; i < arr.length; i++) {
			if (arr[i] < head)
				left.add(arr[i]);
			if (arr[i] > head)
				right.add(arr[i]);
		}

		Collections.sort(left);
		Collections.sort(right);

		if (direction.equals("right")) {
			for (int i = 0; i < right.size(); i++) {
				cur_track = right.get(i);
				seek_sequence.add(cur_track);
				distance = Math.abs(cur_track - head);
				seek_count += distance;
				head = cur_track;
			}

			head = 0;
			seek_count += (disk_size - 1);

			for (int i = 0; i < left.size(); i++) {
				cur_track = left.get(i);
				seek_sequence.add(cur_track);
				distance = Math.abs(cur_track - head);
				seek_count += distance;
				head = cur_track;
			}
		} else if (direction.equals("left")) {
			for (int i = 0; i < left.size(); i++) {
				cur_track = left.get(i);
				seek_sequence.add(cur_track);
				distance = Math.abs(cur_track - head);
				seek_count += distance;
				head = cur_track;
			}

			head = disk_size - 1;
			seek_count += (disk_size - 1);

			for (int i = 0; i < right.size(); i++) {
				cur_track = right.get(i);
				seek_sequence.add(cur_track);
				distance = Math.abs(cur_track - head);
				seek_count += distance;
				head = cur_track;
			}
		} else {
			System.out.println("Invalid direction! Please specify either 'left' or 'right'.");
			return;
		}

		System.out.println("Total number of seek operations = " + seek_count);

		System.out.println("Seek Sequence is");

		for (int i = 0; i < seek_sequence.size(); i++) {
			System.out.print(seek_sequence.get(i));
			if (i != seek_sequence.size() - 1) {
				System.out.print(" -> ");
			}
		}
	}
}

File: c cpp/Disk Scheduling/cscan.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int disk_size;

void CSCAN(vector<int> arr, int head, string direction)
{
    int seek_count = 0;
    int distance, cur_track;
    vector<int> left, right, seek_sequence;

    left.push_back(0);
    right.push_back(disk_size - 1);

    for (int i = 0; i < arr.size(); i++)
    {
        if (arr[i] < head)
            left.push_back(arr[i]);
        if (arr[i] > head)
            right.push_back(arr[i]);
    }

    sort(left.begin(), left.end());
    sort(right.begin(), right.end());

    if (direction == "right")
    {
        for (int i = 0; i < right.size(); i++)
        {
            cur_track = right[i];
            seek_sequence.push_back(cur_track);
            distance = abs(cur_track - head);
            seek_count += distance;
            head = cur_track;
        }

        head = 0;
        seek_count += (disk_size - 1);

        for (int i = 0; i < left.size(); i++)
        {
            cur_track = left[i];
            seek_sequence.push_back(cur_track);
            distance = abs(cur_track - head);
            seek_count += distance;
            head = cur_track;
        }
    }
    else if (direction == "left")
    {
        for (int i = 0; i < left.size(); i++)
        {
            cur_track = left[i];
            seek_sequence.push_back(cur_track);
            distance = abs(cur_track - head);
            seek_count += distance;
            head = cur_track;
        }

        head = disk_size - 1;
        seek_count += (disk_size - 1);

        for (int i = 0; i < right.size(); i++)
        {
            cur_track = right[i];
            seek_sequence.push_back(cur_track);
            distance = abs(cur_track - head);
            seek_count += distance;
            head = cur_track;
        }
    }
    else
    {
        cout << "Invalid direction! Please specify either 'left' or 'right.'" << endl;
        return;
    }

    cout << "Total number of seek operations = " << seek_count << endl;

    cout << "Seek Sequence is" << endl;

    for (int i = 0; i < seek_sequence.size(); i++)
    {
        cout << seek_sequence[i] << endl;
    }
}

int main()
{
    int size, head;
    cout << "Enter the number of requests: ";
    cin >> size;
    vector<int> arr(size);

    cout << "Enter the requests:" << endl;
    for (int i = 0; i < size; i++)
    {
        cin >> arr[i];
    }

    cout << "Enter the initial head position: ";
    cin >> head;

    cout << "Enter the disk size: ";
    cin >> disk_size;

    string direction;
    cout << "Enter the direction of head movement (left or right): ";
    cin >> direction;

    cout << "Initial position of head: " << head << endl;

    CSCAN(arr, head, direction);

    return 0;
}
